src/document-symbols.jl

```diff
@@ -0,0 +1,313 @@
+                          selectionRange = range,
+                          children = children)
+end
+
+# TODO: `end` keyword is not highlighted.
+"""
+Get the `LSP.Range` of a `SyntaxTree`.
+"""
+function get_range(ex::JL.SyntaxTree)
+    (line, col) = JS.source_location(ex)
+    start_position = Position(; line = line - 1, character = col - 1)
+    # Calculate the end position from the byte range and the line start byte.
+    line_starts = JS.sourcefile(ex).line_starts
+    byte_range = JS.byte_range(ex)
+    if length(line_starts) == line || byte_range.stop < line_starts[line + 1]
+        # `ex` does not span multiple lines.
+        # TODO: Calculate the last column in terms of characters, not bytes.
+        end_position = Position(; line = line - 1, character = col + length(byte_range) - 1)
+    else
+        # `ex` spans multiple lines.
+        end_byte = byte_range.stop
+        # TODO: Is there a smarter way to do this?
+        # Calculate the ending line.
+        end_line = line
+        for i in line+1:length(line_starts)-1
+            if line_starts[i + 1] > end_byte
+                end_line = i
+                break
+            end
+        end
+        if end_line == line
+            end_line = length(line_starts)
+        end
+        # Calculate the ending column.
+        # TODO: Calculate it in terms of characters, not bytes.
+        end_col = end_byte - line_starts[end_line]
+        # End position.
+        end_position = Position(; line = end_line - 1, character = end_col - 1)
+    end
+
+    return Range(; start = start_position, var"end" = end_position)
+end
+
+"""
+Get the function name from a function signature node.
+  - `f`
+  - `f(...)`
+  - `f(...)::T`
+"""
+function get_function_name(ex::JL.SyntaxTree)
+    k = kind(ex)
+    if k === K"Identifier"
+        return ex.name_val
+    elseif k === K"call"
+        name_ex = children(ex)[1]
+        return kind(name_ex) === K"Identifier" ?
+            name_ex.name_val :
+            "anonymous"
+    elseif k === K"::"
+        call_ex = children(ex)[1]
+        return get_function_name(call_ex)
+    end
+end
+
+function push_args!(arg_exs, fun_call_ex, start)
+    for arg_ex in fun_call_ex[start:end]
+        kind(arg_ex) === K"parameters" ?
+            push!(arg_exs, arg_ex[1])  :
+            push!(arg_exs, arg_ex)
+    end
+    return arg_exs
+end
+
+"""
+Get the type name from the rhs of a `::` node.
+  - `::T`      -> "T"
+  - `::T1{T2}` -> "T1{T2}"
+"""
+function get_type_name(ex::JL.SyntaxTree)
+    return JS.is_identifier(ex) ?
+        ex.name_val             :
+        ex[1].name_val * "{" * get_type_name(ex[2]) * "}"
+end
+
+# Request handler
+# ===============
+
+function handle_DocumentSymbolRequest(state::ServerState, msg::DocumentSymbolRequest)
+    file_uri = URI(msg.params.textDocument.uri)
+    file_info = get_fileinfo(state, file_uri)
+    isnothing(file_info) &&
+        # TODO: Send error/warning/diagnostic?
```

Comment:
```suggestion
```
No need to emit any diagnostics here. Diagnostics are emitted by the server upon `onTextDocumentChange` notification.